ls1intum/Artemis
# `Development`: Avoid unnecessary REST calls in course details

### Checklist
#### General


#### Client

### Motivation and Context



The learning goals, tutorial groups, and prerequisites of a course are fetched already when calling courses/{courseId}/for-dashboard API. Therefore separate calls are not needed when visiting a course. The API calls are shown in this figure:

![before](https://user-images.githubusercontent.com/10885503/232450992-72db3ae4-a3ec-4f4a-8edc-c39d1d4cf4f2.png)

### Description

1. Remove /courses/{courseId}/for-dashboard API call from MetisConversationService
2. Always call /courses/{courseId}/for-dashboard API from the Course Overview Component and pass the fetched course to MetisConversationService
3. Remove additional API calls to get Learning Goals, Prerequisites and Tutorial Groups

### Steps for Testing

Prerequisites:
- 1 Student
- 1 Course with Learning Goals, Prerequisites, and Tutorial Groups

1. Log in to Artemis as a Student
4. Navigate to Course
5. Open network tabs in browser
6. Select 1 of the Course
7. Verify that there are no additional API calls to retrieve learning goals, prerequisites, and tutorial groups
8. Verify that learning goals, prerequisites and tutorial groups are accessible in the course detail

### Review Progress






#### Code Review
#### Manual Tests

### Test Coverage



| Class/File | Branch | Line |
|------------|-------:|-----:|
| course-overview.component.ts | 75.6% | 85.49% |
| metis-conversation.service.ts | 71.42% | 85% |

### Screenshots



After the changes in this PR, no separate API calls are made:
<img width="594" alt="image" src="https://user-images.githubusercontent.com/10885503/232541044-cbb05573-3fe4-4be4-8519-dfbef6f26b97.png">

Learning Goals and Prerequisites:
<img width="1037" alt="image" src="https://user-images.githubusercontent.com/10885503/232541149-222a7645-31aa-4964-a279-421f996f4ee5.png">

Tutorial Groups:
<img width="1512" alt="image" src="https://user-images.githubusercontent.com/10885503/232541259-58df093e-48a4-4a1c-a9f7-21495479f17c.png">
